docarray/utils/apply.py

```diff
@@ -0,0 +1,93 @@
+from contextlib import nullcontext
+from types import LambdaType
+from typing import TYPE_CHECKING, Any, Callable, Generator, Optional, TypeVar, Union
+
+from docarray import BaseDocument
+from docarray.array.abstract_array import AnyDocumentArray
+
+if TYPE_CHECKING:
+    from multiprocessing.pool import Pool
+
+
+T = TypeVar('T', bound=AnyDocumentArray)
+
+
+def apply(
+    da: T,
+    func: Callable[[BaseDocument], BaseDocument],
+    num_worker: Optional[int] = None,
+    pool: Optional['Pool'] = None,
+    show_progress: bool = False,
```

Comment:
Should we have a `backend` options to switch between multi-processing and multi-threading?

Reply:
@JoanFM maybe know more. But to me only multi processing makes sense. Can multi threading really improve performance here ?

Reply:
Just the keep the PR discussion up to date with what we discussed on Discord: there will be multi threading since it makes sense for IO bound ops and tf/np/torch stuff